libs/datamodel/core/src/transform/ast_to_dml/db/relations.rs

```diff
@@ -0,0 +1,219 @@
+            .map(move |(_model_b_id, model_a_id, relation_idx)| (*model_a_id, &self.relations_storage[*relation_idx]))
+    }
+}
+
+pub(super) fn infer_relations(ctx: &mut Context<'_>) {
+    let mut relations = Relations::default();
+
+    for rf in ctx.db.types.relation_fields.iter() {
+        let evidence = relation_evidence(rf, ctx);
+        ingest_relation(evidence, &mut relations);
+    }
+
+    ctx.db.relations = relations;
+}
+
+// Implementation detail for this module. Should stay private.
+struct RelationEvidence<'ast, 'db> {
+    ast_model: &'ast ast::Model,
+    model_id: ast::ModelId,
+    ast_field: &'ast ast::Field,
+    field_id: ast::FieldId,
+    is_self_relation: bool,
+    relation_field: &'db RelationField<'ast>,
+    opposite_model: &'ast ast::Model,
+    opposite_relation_field: Option<(ast::FieldId, &'ast ast::Field, &'db RelationField<'ast>)>,
+}
+
+fn relation_evidence<'ast, 'db>(
+    ((model_id, field_id), relation_field): (&(ast::ModelId, ast::FieldId), &'db RelationField<'ast>),
+    ctx: &'db Context<'ast>,
+) -> RelationEvidence<'ast, 'db> {
+    let ast_model = &ctx.db.ast[*model_id];
+    let ast_field = &ast_model[*field_id];
+    let opposite_model = &ctx.db.ast[relation_field.referenced_model];
+    let is_self_relation = *model_id == relation_field.referenced_model;
+    let opposite_relation_field: Option<(ast::FieldId, &ast::Field, &RelationField<'_>)> = ctx
+        .db
+        .iter_model_relation_fields(relation_field.referenced_model)
+        // Only considers relations between the same models
+        .filter(|(_, opposite_relation_field)| opposite_relation_field.referenced_model == *model_id)
+        .filter(|(opposite_field_id, _)| !is_self_relation || opposite_field_id != field_id)
+        // Filter out the field itself, in case of self-relations
+        .filter(|(opposite_field_id, _)| !is_self_relation || opposite_field_id != field_id)
+        .find(|(_, opp)| opp.name == relation_field.name)
+        .map(|(opp_field_id, opp)| (opp_field_id, &opposite_model[opp_field_id], opp));
+
+    RelationEvidence {
+        ast_model,
+        model_id: *model_id,
+        ast_field,
+        field_id: *field_id,
+        relation_field,
+        opposite_model,
+        is_self_relation,
+        opposite_relation_field,
+    }
+}
+
+fn ingest_relation<'ast, 'db>(evidence: RelationEvidence<'ast, 'db>, relations: &mut Relations<'ast>) {
+    let relation_type = match (evidence.ast_field.arity, evidence.opposite_relation_field) {
+        (ast::FieldArity::List, Some((opp_field_id, opp_field, _))) if opp_field.arity.is_list() => {
+            // This is an implicit many-to-many relation.
+
+            // We will meet the relation twice when we walk over all relation
+            // fields, so we only instantiate it when the relation field is that
+            // of model A, and the opposite is model B.
+            if evidence.ast_model.name.name > evidence.opposite_model.name.name {
```

Comment:
`name.name` -> `name()`, and please make the `name` property private.

Reply:
I don't think we can (make it private), because we construct `Identifier`'s in a bunch of places. I'll still see if it's possible.